Zend/zend_execute.c

```diff
@@ -795,10 +863,69 @@ static zend_never_inline int zend_verify_internal_arg_types(zend_function *fbc,
 	return 1;
 }
 
+static zend_always_inline zend_bool zend_check_type(
+		const zend_function *zf, const zend_arg_info *arg_info,
+		zval *arg, zend_class_entry **ce, void **cache_slot,
+		zval *default_value, zend_bool is_return_type)
+{
+	if (!arg_info->type_hint) {
+		return 1;
+	}
+
+	ZVAL_DEREF(arg);
+	if (EXPECTED(arg_info->type_hint == Z_TYPE_P(arg))) {
+		if (arg_info->class_name) {
+			if (EXPECTED(*cache_slot)) {
+				*ce = (zend_class_entry *) *cache_slot;
+			} else {
+				*ce = zend_verify_arg_class_kind(arg_info);
+				if (UNEXPECTED(!*ce)) {
+					return 0;
+				}
+				*cache_slot = (void *) *ce;
+			}
+			if (UNEXPECTED(!instanceof_function(Z_OBJCE_P(arg), *ce))) {
+				return 0;
+			}
+		}
+		return 1;
+	}
+
+	if (Z_TYPE_P(arg) == IS_NULL && (arg_info->allow_null || (default_value && is_null_constant(zf->common.scope, default_value)))) {
```

Comment:
This handles cases like the following:

```php
function test(Foo $bar = ABC) {}
define('ABC', null);
test(null);
```

Here we cannot determine at compile-time that the default value is actually null.

So yeah, at least theoretically the "or null" handling would have to be adjusted for this. However, I feel like we should instead drop this special case altogether. IIRC it was introduced as a bug "fix" sometime in 5.6 without anyone properly evaluating the consequences. For example, this means that `allowsNull()` in reflection lies. It also means that our LSP checks do not work correctly. Now that PHP 7.1 has an explicit way to mark that argument as nullable using `?Foo $bar = ABC`, this should be dropped.